### ISSUE TITLE
Mqtt payload for publish message copied 2 times during lifetime

### ISSUE THREAD

**Issue by doom369** [netty/netty#13844]
### PublishBuilder.build():

```
        public MqttPublishMessage build() {
            MqttFixedHeader mqttFixedHeader = new MqttFixedHeader(MqttMessageType.PUBLISH, false, qos, retained, 0);
            MqttPublishVariableHeader mqttVariableHeader =
                    new MqttPublishVariableHeader(topic, messageId, mqttProperties);
            return new MqttPublishMessage(mqttFixedHeader, mqttVariableHeader, Unpooled.buffer().writeBytes(payload));
        }
```

has `Unpooled.buffer().writeBytes(payload)`. This is one is not big deal as we can easily overcome it.

### MqttEncoder.encodePublishMessage:

`ByteBuf payload = message.payload().duplicate();`

My understanding, the authors were playing safe, however 2 copies, seems like overkill. WDYT? Should I fix it?

**Comment by normanmaurer** [comment]
@doom369 sounds like something we can change .. Do you want to do a PR ?

I think it should be `message.payload().retainedDuplicate()`

**Comment by doom369** [comment]
@normanmaurer I was thinking about:

`message.payload().duplicate()` -> `message.payload()`

In a typical use case you would do:

```
        ByteBuf buffer = ctx.alloc().buffer(body.length());
        buffer.writeCharSequence(body, StandardCharsets.UTF_8);

        MqttPublishMessage message = MqttMessageBuilders.publish()
                .payload(buffer)
                .qos(MqttQoS.AT_LEAST_ONCE)
                .messageId(packetId)
                .build();
       ctx.write(message);
```

Do you think this `duplicate()` has some purpose?

**Comment by franz1981** [comment]
I would allocate an ioBuffer and perform the copy into it, which will still happen later in the transport layers...

**Comment by doom369** [comment]
> we could have a builder which allow to pass some arbitrary buffer too, and will allow users to allocate it themselves

That's what the current `MqttMessageBuilders` has already. The question - of why the encoder does duplicate(), doesn't seem necessary for me.

**Comment by lhotari** [comment]
> Unpooled.buffer().writeBytes(payload)

One reason for extra copies in Apache Pulsar is due to the SslHandler "feature" of mutating the input buffers in some cases. The problem with the use case is reported as #14069 (remaining parts addressed by PR #14086). Perhaps this isn't the case here.

> ByteBuf payload = message.payload().duplicate();

> That's what the current MqttMessageBuilders has already. The question - why the encoder does duplicate(), doesn't seem necessary for me.

this isn't a real issue since the reason to use `.duplicate()` is to ensure that the payload buffer's readerIndex, writerIndex and markers don't get modified. The implementation is optimized and doesn't copy the buffer's content.
It would be a breaking change to remove the `.duplicate()` call.

**Comment by doom369** [comment]
Not big deal. Fixed in own code.
